apps/prairielearn/src/lib/server-jobs-2.ts

```diff
@@ -0,0 +1,184 @@
+import AnsiUp from 'ansi_up';
+import execa = require('execa');
+import { z } from 'zod';
+import * as Sentry from '@prairielearn/sentry';
+import { logger } from '@prairielearn/logger';
+import { loadSqlEquiv, queryAsync, queryValidatedOneRow } from '@prairielearn/postgres';
+
+import { chalk } from './chalk';
+import serverJobs = require('./server-jobs');
+import socketServer = require('./socket-server');
+
+const sql = loadSqlEquiv(__filename);
+
+interface CreateServerJobOptions {
+  courseId?: string;
+  courseInstanceId?: string;
+  courseRequestId?: string;
+  assessmentId?: string;
+  userId?: string;
+  authnUserId?: string;
+  type: string;
+  description: string;
+}
+
+interface ServerJobExecOptions {
+  cwd?: string;
+  env?: NodeJS.ProcessEnv;
+}
+
+type ServerJobContext = Omit<ServerJob, 'execute' | 'executeInBackground'>;
+type JobExecutionFunction = (context: ServerJobContext) => Promise<void>;
+
+class ServerJob {
+  public jobSequenceId: string;
+  public jobId: string;
+  private started = false;
+  private finished = false;
+  public output = '';
+
+  constructor(jobSequenceId: string, jobId: string) {
+    this.jobSequenceId = jobSequenceId;
+    this.jobId = jobId;
+  }
+
+  error(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  warn(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  info(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  verbose(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  async exec(file: string, args: string[] = [], options: ServerJobExecOptions = {}): Promise<void> {
+    const msg = `$ ${file} ${args.join(' ')}\n`;
```

Comment:
Should we be logging the working directory somehow? We previously persisted that information in `jobs.working_directory`.

Reply:
I don't know that we want to display this publicly, but it feels like it could be helpful for debugging (like the environment for the job). Is it worth adding a secret log for these that we persist into the DB?

Reply:
Actually I think we used to display it publicly so we probably could still do so if that's the easiest option?

Reply:
In Alma I render commands as `cd /path/to/working/directory && git command --goes-here`. I could also do something like `/path/to/working/directory $ git command --goes-here`. Or even this:

```
Working directory: /path/to/working/directory
Command: git command --goes-here
```

Any preference?

Reply:
Either the `/dir $ cmd` or the more verbose syntax look good to me. If I was forced to pick then I might pick the second one to avoid super long lines, but it's honestly a toss-up.

Reply:
I went with the verbose one in 35f7710.